Source/WebKit/Shared/graphics/ImageBufferSet.cpp

```diff
@@ -0,0 +1,153 @@
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "ImageBufferSet.h"
+
+#include <WebCore/GraphicsContext.h>
+#include <WebCore/ImageBufferBackend.h>
+
+#if PLATFORM(COCOA)
+#include <WebCore/PlatformCALayer.h>
+#endif
+
+#if ENABLE(GPU_PROCESS)
+
+namespace WebKit {
+
+using namespace WebCore;
+
+SwapBuffersDisplayRequirement ImageBufferSet::swapBuffersForDisplay(bool hasEmptyDirtyRegion, bool supportsPartialRepaint)
+{
+    auto displayRequirement = SwapBuffersDisplayRequirement::NeedsNoDisplay;
+
+    // Make the previous front buffer non-volatile early, so that we can dirty the whole layer if it comes back empty.
+    if (!m_frontBuffer || m_frontBuffer->setNonVolatile() == WebCore::SetNonVolatileResult::Empty)
+        displayRequirement = SwapBuffersDisplayRequirement::NeedsFullDisplay;
+    else if (!hasEmptyDirtyRegion)
+        displayRequirement = SwapBuffersDisplayRequirement::NeedsNormalDisplay;
+
+    if (displayRequirement == SwapBuffersDisplayRequirement::NeedsNoDisplay)
+        return displayRequirement;
+
+    if (!supportsPartialRepaint)
+        displayRequirement = SwapBuffersDisplayRequirement::NeedsFullDisplay;
+
+    if (!m_backBuffer || m_backBuffer->isInUse()) {
+        m_previouslyPaintedRect = std::nullopt;
+        std::swap(m_backBuffer, m_secondaryBackBuffer);
+
+        // When pulling the secondary back buffer out of hibernation (to become
+        // the new front buffer), if it is somehow still in use (e.g. we got
+        // three swaps ahead of the render server), just give up and discard it.
+        if (m_backBuffer && m_backBuffer->isInUse())
+            m_backBuffer = nullptr;
+    }
+
+    std::swap(m_frontBuffer, m_backBuffer);
+
+    if (m_frontBuffer) {
+        if (m_frontBuffer->setNonVolatile() == WebCore::SetNonVolatileResult::Empty)
+            m_previouslyPaintedRect = std::nullopt;
+    }
+
+    return displayRequirement;
+}
+
+Vector<FloatRect, 5> ImageBufferSet::computePaintingRects(const Region& dirtyRegion, float resolutionScale)
```

Comment:
Should the magic number 5 be a named constant and/or Vector<FloatRect, 5> have a `using` alias? (eg. `using PaintRects = Vector<FloatRect, 5>`)